src/ralph/backends/data/async_ws.py

```diff
@@ -0,0 +1,186 @@
+            bytes: If `raw_output` is True.
+
+        Raise:
+            BackendException: If a failure during the read operation occurs or
+                during encoding records and `ignore_errors` is set to `False`.
+            BackendParameterException: If a backend argument value is not valid.
+        """
+        statements = super().read(
+            query,
+            target,
+            chunk_size,
+            raw_output,
+            ignore_errors,
+            prefetch,
+            max_statements,
+        )
+        async for statement in statements:
+            yield statement
+
+    async def _read_bytes(
+        self,
+        query: BaseQuery,  # noqa: ARG002
+        target: Optional[str],
+        chunk_size: int,  # noqa: ARG002
+        ignore_errors: bool,  # noqa: ARG002
+    ) -> AsyncIterator[bytes]:
+        """Method called by `self.read` yielding bytes. See `self.read`."""
+        uri = target if target else self.settings.URI
+        client = await self.client(uri)
+        count = 0
+        try:
+            while event := await client.recv():
+                yield bytes(f"{event}" + "\n", encoding=self.settings.LOCALE_ENCODING)
+                count += 1
+        except websockets.exceptions.ConnectionClosedOK:
+            self.logger.info("Read %s records with success", count)
+        except (websockets.WebSocketException, RuntimeError) as error:
+            msg = "Failed to receive message from websocket %s: %s"
+            self.logger.error(msg, uri, error)
+            raise BackendException(msg % (uri, error)) from error
+
+    async def _read_dicts(
+        self,
+        query: BaseQuery,
+        target: Optional[str],
+        chunk_size: int,
+        ignore_errors: bool,
+    ) -> AsyncIterator[dict]:
+        """Method called by `self.read` yielding dictionaries. See `self.read`."""
+        statements = self._read_bytes(query, target, chunk_size, ignore_errors)
+        statements = async_parse_iterable_to_dict(
+            statements, ignore_errors, self.logger
+        )
+        async for statement in statements:
+            yield statement
+
+    async def close(self) -> None:
+        """Close the websocket client.
+
+        Raise:
+            BackendException: If a failure occurs during the close operation.
+        """
+        if not self._clients:
+            self.logger.warning("No backend client to close.")
+            return
+
+        errors = {}
+        for uri, client in self._clients.items():
+            try:
+                await client.close()
+            except (websockets.WebSocketException, OSError, TimeoutError) as error:
+                msg = "Failed to close websocket connection for %s: %s"
+                self.logger.error(msg, uri, error)
+                errors[uri] = error
+
+        self._clients = {}
+        if errors:
+            msg = "Failed to close websocket connections: %s"
+            raise BackendException(msg % errors)
```

Comment:
I'd rather raise an exception earlier when the first connection closing fails.

Reply:
Good point) Why should we keep the remaining connections open if one fails? Wouldn't it be problematic if the caller catches the `BackendException` and continues execution?

Reply:
Dropping multi-ws client support simplified this)